include/sparrow/mp_utils.hpp

```diff
@@ -14,19 +14,228 @@
+        ///     static_assert( all_of(some_types, same_as_int) == false);
+        template<class T>
+        struct same_as
+        {
+            template<class X>
+            consteval
+            bool operator()(mpl::typelist<X> list) const
+            {
+                return std::same_as< mpl::typelist<T>, decltype(list) >;
+            }
+
+        };
+    }
+
+    template< template<class> class P >
+    struct ct_type_predicate_to_callable
     {
-        template <class T, bool is_const>
-        struct constify
-            : std::conditional<is_const, const T, T>
+        template<template<class> class W, class T>
+            requires ct_type_predicate<P, T> and type_wrapper<W,T>
+        consteval bool operator()(W<T>)
         {
+            return P<T>::value;
+        }
+    };
+
+
+    template< template<class> class P >
+    consteval
+    auto as_predicate()
+    {
+        return ct_type_predicate_to_callable<P>{};
+    };
+
+
+
+    //////////////////////////////////////////////////
+    //// Algorithms //////////////////////////////////
+
+    /// Checks that at least one type in the provided list of is making the provide predicate return `true`.
+    /// @returns 'true' if for at least one type T in the type list L,, `Predicate{}(typelist<T>) == true`.
+    ///          `false` otherwise or if the list is empty.
+    template< class Predicate, template<class...> class L, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+    bool any_of(L<T...> list, Predicate predicate = {})
+    {
+        return (evaluate<T>(predicate) || ... || false);
+    }
+
+    template< template<class> class Predicate, template<class...> class L, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+    bool any_of(L<T...> list)
+    {
+        return any_of(list, as_predicate<Predicate>());
+    }
+
+    /// Checks that every type in the provided list of is making the provide predicate return `true`.
+    /// @returns `true` if for every type T in the type list L, `Predicate{}(typelist<T>) == true`
+    ///          or if the list is empty; `false` otherwise.
+    template< template<class...> class L, class Predicate, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+    bool all_of(L<T...> list, Predicate predicate)
+    {
+        return (evaluate<T>(predicate) && ... && true);
+    }
+
+
+    /// @returns `true` if the provided type list contains `V`
+    template< class V, any_typelist L>
+    consteval
+    bool contains(L list)
+    {
+        return any_of(list, predicate::same_as<V>{});
+    }
+
+
+    /// @returns The index position in the first type in the provided type list `L` that matches the provided predicate,
+    ///          or the size of the list if the matching type was not found.
+    template< template<class...> class L, class Predicate, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+    std::size_t find_if(L<T...> list, Predicate predicate)
```

Comment:
This function has an interesting implementation.

Reply:
Yeah I even sent the code to @JohanMabille ask if it looked weird XD